plugin/pkg/scheduler/core/generic_scheduler.go

```diff
@@ -318,34 +376,69 @@ func podFitsOnNode(pod *v1.Pod, meta algorithm.PredicateMetadata, info *schedule
 		reasons          []algorithm.PredicateFailureReason
 		err              error
 	)
+	predicateResults := make(map[string]HostPredicate)
+
 	if ecache != nil {
 		// getHashEquivalencePod will return immediately if no equivalence pod found
 		equivalenceHash, eCacheAvailable = ecache.getHashEquivalencePod(pod)
 	}
-	for predicateKey, predicate := range predicateFuncs {
-		// If equivalenceCache is available
-		if eCacheAvailable {
-			// PredicateWithECache will returns it's cached predicate results
-			fit, reasons, invalid = ecache.PredicateWithECache(pod.GetName(), info.Node().GetName(), predicateKey, equivalenceHash)
-		}
-
-		if !eCacheAvailable || invalid {
-			// we need to execute predicate functions since equivalence cache does not work
-			fit, reasons, err = predicate(pod, meta, info)
-			if err != nil {
-				return false, []algorithm.PredicateFailureReason{}, err
+	podsAdded := false
+	// We run predicates twice. Once we run them when pods nominated to run on
+	// the node added to meta and nodeInfo. If there are such pods and if all
+	// predicates succeed in this pass, we run them again when these nominated pods
+	// are not added. This second pass is necessary because some predicates such
+	// as inter-pod affinity may not pass without the nominated pods.
+	for i := 0; i < 2; i++ {
+		metaToUse := meta
+		nodeInfoToUse := info
+		if i == 0 {
+			podsAdded, metaToUse, nodeInfoToUse = addNominatedPods(util.GetPodPriority(pod), meta, info, queue)
+		} else if !podsAdded || len(failedPredicates) != 0 {
```

Comment:
IIUC this "else if" clause is enforcing the following:
* If there are no nominated pods for this node, then use the feasibility result from the previous iteration of the loop, which tested the pod against the node with no nominated pods (of course)
* If there are nominated pods for this node, and the pod is infeasible with the nominated pods added (which was determined on the previous iteration of the loop), then declare the pod to be infeasible

Is that right?

If so, I am wondering about the second thing.

What is the high-level rule you are trying to enforce? It seems like a reasonable rule might be "The pod is feasible if it will be feasible once all of the victims are gone and the preemptors have started running." But IIUC the rule you are enforcing is "The pod is feasible if it is feasible with all preemptors and victims running simultaneously AND it is feasible in the current state (victims running, preemptors not running yet)." Can you explain the reason behind this rule?

Reply:
> IIUC this "else if" clause is enforcing the following:
>If there are no nominated pods for this node, then use the feasibility result from the previous iteration of the loop, which tested the pod against the node with no nominated pods (of course)
If there are nominated pods for this node, and the pod is infeasible with the nominated pods added (which was determined on the previous iteration of the loop), then declare the pod to be infeasible
Is that right?

Yes, your understanding is right.

> What is the high-level rule you are trying to enforce? It seems like a reasonable rule might be "The pod is feasible if it will be feasible once all of the victims are gone and the preemptors have started running." But IIUC the rule you are enforcing is "The pod is feasible if it is feasible with all preemptors and victims running simultaneously AND it is feasible in the current state (victims running, preemptors not running yet)." Can you explain the reason behind this rule? This seems to be testing a condition that will never happen (victims and preemptors running simultaneously). I understand you care about the preemptors because of pod affinity but IIUC you are testing against a configuration of the node that will never actually exist. Would testing just the pod affinity predicate help? (Though I'm not sure what configuration you'd use it with.)

Please note that this function is called from two different places: Schedule and Preempt.
When it is called in Schedule, we want to test whether the pod is schedulable on the node with all the existing pods on the node plus higher (and equal) priority pods nominated to run on the node.
When it is called from Preempt, we should do what you said. We should remove the victims and add the nominated pods. Removal of the victims is done by [SelectVictimsOnNode()](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/core/generic_scheduler.go#L693). It removes victims from `meta` and `NodeInfo` before calling this function.

Reply:
Did you add a comment with this (at the top of this function)? If not, please do

> this function is called from two different places: Schedule and Preempt.
When it is called in Schedule, we want to test whether the pod is schedulable on the node with all the existing pods on the node plus higher (and equal) priority pods nominated to run on the node.
When it is called from Preempt, we should do what you said. We should remove the victims and add the nominated pods. Removal of the victims is done by SelectVictimsOnNode(). It removes victims from meta and NodeInfo before calling this function.

Reply:
Done